jetstream_consumer.go

```diff
@@ -0,0 +1,335 @@
+func (e ApiError) NotFoundError() bool { return e.Code == 404 }
+
+// ServerError is true when the server returns a 5xx error code
+func (e ApiError) ServerError() bool { return e.Code >= 500 && e.Code < 600 }
+
+// UserError is true when the server returns a 4xx error code
+func (e ApiError) UserError() bool { return e.Code >= 400 && e.Code < 500 }
+
+// ErrorCode is the JetStream error code
+func (e ApiError) ErrorCode() int { return e.Code }
+
+type JSApiResponse struct {
+	Type  string    `json:"type"`
+	Error *ApiError `json:"error,omitempty"`
+}
+
+// ToError extracts a standard error from a JetStream response
+func (r JSApiResponse) ToError() error {
+	if r.Error == nil {
+		return nil
+	}
+
+	return *r.Error
+}
+
+// IsError determines if a standard JetStream API response is a error
+func (r JSApiResponse) IsError() bool {
+	return r.Error != nil
+}
+
+// io.nats.jetstream.api.v1.consumer_create_request
+type JSApiConsumerCreateRequest struct {
+	Stream string         `json:"stream_name"`
+	Config ConsumerConfig `json:"config"`
+}
+
+// io.nats.jetstream.api.v1.consumer_create_response
+type JSApiConsumerCreateResponse struct {
+	JSApiResponse
+	*ConsumerInfo
+}
+
+type AckPolicy int
+
+const (
+	AckNone AckPolicy = iota
+	AckAll
+	AckExplicit
+)
+
+func (p AckPolicy) String() string {
+	switch p {
+	case AckNone:
+		return "None"
+	case AckAll:
+		return "All"
+	case AckExplicit:
+		return "Explicit"
+	default:
+		return "Unknown Acknowledgement Policy"
+	}
+}
+
+func (p *AckPolicy) UnmarshalJSON(data []byte) error {
+	switch string(data) {
+	case jsonString("none"):
+		*p = AckNone
+	case jsonString("all"):
+		*p = AckAll
+	case jsonString("explicit"):
+		*p = AckExplicit
+	default:
+		return fmt.Errorf("can not unmarshal %q", data)
+	}
+
+	return nil
+}
+
+func (p AckPolicy) MarshalJSON() ([]byte, error) {
+	switch p {
+	case AckNone:
+		return json.Marshal("none")
+	case AckAll:
+		return json.Marshal("all")
+	case AckExplicit:
+		return json.Marshal("explicit")
+	default:
+		return nil, fmt.Errorf("unknown acknowlegement policy %v", p)
+	}
+}
+
+type ReplayPolicy int
+
+const (
+	ReplayInstant ReplayPolicy = iota
+	ReplayOriginal
+)
+
+func (p ReplayPolicy) String() string {
+	switch p {
+	case ReplayInstant:
+		return "Instant"
+	case ReplayOriginal:
+		return "Original"
+	default:
+		return "Unknown Replay Policy"
+	}
+}
+
+func (p *ReplayPolicy) UnmarshalJSON(data []byte) error {
+	switch string(data) {
+	case jsonString("instant"):
+		*p = ReplayInstant
+	case jsonString("original"):
+		*p = ReplayOriginal
+	default:
+		return fmt.Errorf("can not unmarshal %q", data)
+	}
+
+	return nil
+}
+
+func (p ReplayPolicy) MarshalJSON() ([]byte, error) {
+	switch p {
+	case ReplayOriginal:
+		return json.Marshal("original")
+	case ReplayInstant:
+		return json.Marshal("instant")
+	default:
+		return nil, fmt.Errorf("unknown replay policy %v", p)
+	}
+}
+
+var (
+	AckAck      = []byte("+ACK")
```

Comment:
"+ACK" came on my radar a couple of days ago. IMHO the regular Ack should just be the empty payload. The other values are options to the response which alter the meaning.

Reply:
Used to be, Derek prefers +ACK both are accepted.  Regardless, copied from the server so will resolve.